libtenzir/builtins/operators/buffer.cpp

```diff
@@ -0,0 +1,408 @@
+class write_buffer_operator final
+  : public crtp_operator<write_buffer_operator> {
+public:
+  write_buffer_operator() = default;
+
+  explicit write_buffer_operator(uuid id) : id_{id} {
+  }
+
+  auto
+  operator()(generator<table_slice> input, operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    // The internal-write-buffer operator is spawned after the
+    // internal-read-buffer operator, so we can safely get the buffer actor here
+    // after the first yield and then just remove it from the registry again.
+    co_yield {};
+    auto buffer = ctrl.self().system().registry().get<buffer_actor>(
+      fmt::format("tenzir.buffer.{}", id_));
+    TENZIR_ASSERT(buffer);
+    ctrl.self().link_to(buffer);
+    ctrl.self().system().registry().erase(buffer->id());
+    // Now, all we need to do is send our inputs to the buffer batch by batch.
+    for (auto&& events : input) {
+      if (events.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      ctrl.set_waiting(true);
+      ctrl.self()
+        .request(buffer, caf::infinite, atom::write_v, std::move(events))
+        .then(
+          [&]() {
+            ctrl.set_waiting(false);
+          },
+          [&](caf::error& err) {
+            diagnostic::error(err)
+              .note("failed to write to buffer")
+              .emit(ctrl.diagnostics());
+          });
+      co_yield {};
+    }
+  }
+
+  auto name() const -> std::string override {
+    return "internal-write-buffer";
+  }
+
+  auto optimize(expression const& filter, event_order order) const
+    -> optimize_result override {
+    return optimize_result{filter, order, copy()};
+  }
+
+  auto infer_type_impl(operator_type input) const
+    -> caf::expected<operator_type> override {
+    if (input.is<table_slice>()) {
+      return tag_v<table_slice>;
+    }
+    return diagnostic::error("`buffer` does not accept {} as input",
+                             operator_type_name(input))
+      .to_error();
+  }
+
+  friend auto inspect(auto& f, write_buffer_operator& x) -> bool {
+    return f.object(x).fields(f.field("id", x.id_));
+  }
+
+private:
+  uuid id_ = {};
+};
+
+class read_buffer_operator final : public crtp_operator<read_buffer_operator> {
+public:
+  read_buffer_operator() = default;
+
+  explicit read_buffer_operator(uuid id, located<uint64_t> capacity,
+                                std::optional<buffer_policy> policy)
+    : id_{id}, capacity_{capacity}, policy_{policy} {
+  }
+
+  auto policy(operator_control_plane& ctrl) const -> buffer_policy {
+    return policy_.value_or(ctrl.is_hidden() ? buffer_policy::drop
+                                             : buffer_policy::block);
+  }
+
+  auto metrics(operator_control_plane& ctrl) const -> metric_handler {
+    return ctrl.metrics(type{
+      "tenzir.metrics.buffer",
+      record_type{
+        {"used", uint64_type{}},
+        {"remaining", uint64_type{}},
+        {"dropped", uint64_type{}},
+      },
+    });
+  }
```

Comment:
I don't see any metrics. Tested with

```
source {} 
repeat 
buffer 1, policy="drop" 
head 10
```

Reply:
I'll just fix that in this PR, shouldn't be too hard.

Reply:
ac05af6f5bc